### PR TITLE
tracks all outputs including TBPTT and multiple optimizers

### PR DESCRIPTION
**Pull request by williamFalcon**
Enables behavior that was previously not possible:

- passes the outputs of every time step back to the user for epoch_end hooks and for each optimizer in training.
- also enables auto-reduce and auto-padding when using a Result object

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -592,17 +627,33 @@ def run_training_epoch_end(self, epoch_output, checkpoint_accumulator, early_sto
...
+        # determine if using a result obj
+        # ------------------------
+        # [optimizer_idx][training_step_idx][tbptt_index]
+        opt_idx_outputs = epoch_output[0]
+
+        try:
+            sample_obj = opt_idx_outputs[0][0] if isinstance(opt_idx_outputs[0], list) else opt_idx_outputs[0]
+            is_result_obj = len(epoch_output) > 0 and isinstance(sample_obj, Result)
+        except IndexError as e:
+            is_result_obj = False
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:640]: William, I believe this can be improved. The problem I see here is that, if anywhere else, for example in Result, an IndexError occurs then it will propagate to here and be treated wrongly. I highly recommend not to use exception handling for these types of control flow problems, as it can make debugging a lot more painful. See also #2266
